components/ArticleItemWidget/ArticleItemWidget.js

```diff
@@ -0,0 +1,255 @@
+  resetWidgetUI = () => {
+    this.setState(Object.assign({}, initialState, this.initialTagsState));
+  };
+
+  openBar = () => {
+    this.setState({
+      barOpen: true,
+      readHover: true, //for mobile, explain in onTouchMoveTip function
+      touchTarget: 'read',
+    });
+  };
+
+  hoverTag = event => {
+    const name = event.currentTarget.attributes['name'].value;
+    this.setState({
+      [`${name}Hover`]: true,
+    });
+  };
+
+  unHoverTag = event => {
+    const name = event.currentTarget.attributes['name'].value;
+    this.setState({
+      [`${name}Hover`]: false,
+    });
+  };
+
+  /**
+   * since the touch start event can only trigger when the finger down at the first tag,
+   * so it's hard to done with like mouse enter/leave strategy
+   * then I use onTouchMove to dynamic check which tag was touch over
+   */
+  onTouchMoveTip = event => {
+    var changedTouch = event.changedTouches[0];
+    var elem = document.elementFromPoint(
+      changedTouch.clientX,
+      changedTouch.clientY
+    );
+    try {
+      var target_name = elem.attributes['name'].value;
+      target_name &&
+        this.setState(
+          Object.assign({}, this.initialTagsState, {
+            [`${target_name}Hover`]: true,
+            touchTarget: target_name,
+          })
+        );
+    } catch (error) {
+      // touch move not on the widget tag
+    }
+  };
+
+  // trigger the onClick function in tags config, when finger touch end with tag elm
+  onBarTouchEnd = event => {
+    this.preventAll(event);
+    const target = this.tags.find(select => {
+      return select.name === this.state.touchTarget;
+    });
+    target && target.onClick();
+  };
+
+  onMouseLeave = () => {
+    this.closeNotArticleSelection();
+  };
+
+  closeNotArticleSelection = () => {
+    this.setState({
+      barOpen: false,
+      notArticleSelectionDisplay: false,
+    });
+  };
+
+  handleRead = () => {
+    const {
+      id,
+      handleLocalEditorHelperList,
+      read,
+      notArticleReplied,
+    } = this.props;
+    handleLocalEditorHelperList({
+      id,
+      read: !read,
+      notArticleReplied,
+    });
+  };
+
+  handleNotArticleReply = type => {
+    const { id, handleLocalEditorHelperList, read } = this.props;
+    handleLocalEditorHelperList({
+      id,
+      read,
+      notArticleReplied: type,
+    });
+  };
+
+  render() {
+    const { barOpen, notArticleSelectionDisplay } = this.state;
+    return (
+      <ul
+        className={cx('bar', {
+          active: barOpen,
+        })}
+        ref={widget => (this.refWidget = widget)}
+        style={{ zIndex: notArticleSelectionDisplay ? 2 : 1 }} //or the extend component will cover by next ul
+        onClick={this.preventAll}
+        onMouseEnter={this.openBar}
+        onMouseLeave={this.onMouseLeave}
+        onTouchStart={this.openBar}
+        onTouchMove={this.onTouchMoveTip}
+        onTouchEnd={this.onBarTouchEnd}
+      >
+        {this.tags.map((tag, index) => {
```

Comment:
Personally I would like more declarative style, like:

```
<ul>
  <WidgetItem
    description="已讀"
    color="#cfcfcf"
    hover={this.state.hoverIdx === 0}
    index={0}
    onClick={this.handleReadClick}
    onMouseEnter={this.hoverTag}
    onMouseLeave={this.unHoverTag}
  />
  <WidgetItem
    description="非查證"
    color="orange"
    hover={this.state.hoverIdx === 1}
    index={1}
    onClick={this.handleNotArticleClick}
    onMouseEnter={this.hoverTag}
    onMouseLeave={this.unHoverTag}
  >
    {this.state.notArticleSelectionDisplay && <NotArticleExtendSelection replyToNotArticle={this.replyToNotArticle} />}
  </WidgetItem>
</ul>
```

So that we don't need `this.tags`, which is a bit imperative.

Reply:
I think I use the map function to deal with this only for get the index, then feed it to CSS variable. but obviously, it's not really necessary~